### PR TITLE
Support to update search attributes of a schedule

### PR DESCRIPTION
**Pull request by rodrigozhou**
## What changed?

Support to update search attributes of a schedule

## Why?

https://github.com/temporalio/temporal/issues/4858

## How did you test it?

Added tests.

## Potential risks

No.

## Is hotfix candidate?

No

### CODE REVIEW

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:150]: what are these for? we shouldn't be accessing this stuff from the workflow (or the activities either I think)

Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:150]: Forgot to remove.

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1008]: maybe add a comment that we explicitly want to support non-nil `searchAttributes` with empty map to mean "unset all"

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1022]: I think we can filter these out in frontend and assume they're all good here

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1016]: should we return instead of continue?

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1012]: factor out this code so it works for memo also?

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1027]: this is going to fail on keyword list, right? do we have to use `reflect.DeepEqual` here?

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1034]: this didn't seem to disable the lint check, btw

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1071]: are you 100% sure that this is compatible with the previous code? note that if the previous code decided to do an upsert and this does not, or vice versa, then it's a determinism violation. if there's any doubt I'd prefer to just leave the old code

File: service/worker/scheduler/workflow_test.go
Comment (dnr) [inline service/worker/scheduler/workflow_test.go:1591]: test that this got upserted?

File: tests/schedule.go
Comment (dnr) [inline tests/schedule.go:262]: remove this instead?

File: tests/schedule.go
Comment (dnr) [inline tests/schedule.go:412]: do we have to wait for one run? can we just do Eventually on DescribeSchedule and wait for the search attributes?

File: tests/schedule.go
Comment (dnr) [inline tests/schedule.go:449]: same here

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1022]: We cannot. `workflow.GetTypedSearchAttributes` get the search attributes directly from the history events.
cc: @cretz

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1012]: Memo field names are different. We can't refactor this.

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1027]: Good call. Let me fix this.

File: service/worker/scheduler/workflow_test.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow_test.go:1591]: I modified the functional tests to check. I don't know how to get the schedule workflow search attributes in this test framework.

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1071]: Gonna revert.

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1022]: I'm actually changing to use the untyped version for backwards compatibility.

File: .gitmodules
Comment (rodrigozhou) [reply .gitmodules:4]: Will do before merging the PR

Comment (dnr) [reply .gitmodules:4]: I thought the api change was merged already? anyway, that's fine

File: service/worker/scheduler/workflow_test.go
Comment (dnr) [reply service/worker/scheduler/workflow_test.go:1591]: I think you can use OnUpsertSearchAttributes. ideally it would be good to test all the branches of that loop/conditional: if an attribute is present and unchanged, present and changed, present and deleted, not present and set.

File: tests/schedule.go
Comment (dnr) [inline tests/schedule.go:133]: btw, I left this here a while ago.. we should probably do this for real now, like run the test twice, once with "real" search attributes and once with mapped ones. can be in a followup PR

File: tests/schedule.go
Comment (dnr) [inline tests/schedule.go:413]: oh wow, I didn't know about EventuallyWithT. that's so much nicer than the bool version

File: service/worker/scheduler/workflow.go
Comment (dnr) [reply service/worker/scheduler/workflow.go]: oh, I read this wrong, I thought this loop was over the new ones

File: service/worker/scheduler/workflow.go
Comment (dnr) [inline service/worker/scheduler/workflow.go:1020]: possible determinism concern: if this returns false in one version, then we add a new attribute to the reserved set and this returns true, it may upsert something where it didn't before and break on replay. what do you think? I really don't want to use a sideeffect just for this... should we just call upsert with something we know is reserved and let it fail on the server?

File: tests/schedule.go
Comment (rodrigozhou) [reply tests/schedule.go:133]: Actually, the current test setup is already using aliases.

File: service/worker/scheduler/workflow_test.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow_test.go:1591]: I cannot. Because of how the test is setup with `runAcrossContinue`, I would only be able to add a `OnUpsertSearchAttribute.Maybe`, which kind of defeats the purpose here because we won't know for sure if it ran or not.

File: service/worker/scheduler/workflow.go
Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1020]: Actually, the server doesn't fail `UpsertSearchAttributes` for all reserved search attributes. It fails only for system ones. If the user tries to write to `BuildIds` for example, it would work...

Comment (rodrigozhou) [reply service/worker/scheduler/workflow.go:1020]: I think this is more of a general issue in the Server. Basically, if a custom search attribute name becomes a reserved search attribute (system, or predefined), the Server behavior is not well-defined, and likely to break the workflow if it needs to be replayed.

File: .gitmodules
Comment (rodrigozhou) [reply .gitmodules:4]: I'll merge after this PR is approved.

File: service/worker/scheduler/workflow.go
Comment (dnr) [reply service/worker/scheduler/workflow.go:1029]: Bleh :(  Ok, I guess we can't fix it all here. This should be rare enough (and we can write release notes) to say that users need to manually recreate schedules if they use a search attribute that becomes a reserved one in a new release.

Can you write a comment here saying that this is a deliberate determinism violation and explain when it may break and what we/users should do when it (inevitably) happens?

**Review by dnr — COMMENTED** [review]
would you mind adding commits on top / merging instead of force push during a PR, so it's easy to see the diff across revisions?

**Review by dnr — APPROVED** [review]

**Pull request merged**
